src/app/pages/blog/article/article.js

```diff
@@ -1,130 +1,159 @@
+    return Math.round(100 * visibleHeight / totalHeight);
+}
+
+function ArticleBody({bodyData, setReadTime, setBodyRef}) {
+    const bodyRef = useRef();
 
-    get readTime() {
-        const div = this.el.querySelector('.body');
+    useEffect(() => {
+        const div = bodyRef.current;
         const words = div.textContent.split(/\W+/);
         const WORDS_PER_MINUTE = 225;
 
-        return Math.round(words.length / WORDS_PER_MINUTE);
-    }
+        setReadTime(Math.round(words.length / WORDS_PER_MINUTE));
+    }, [bodyData]);
+    useEffect(() => setBodyRef(bodyRef), []);
 
-    onLoaded() {
-        if (super.onLoaded) {
-            super.onLoaded();
-        }
-        if (this.pageData) {
-            this.attachUnits();
-        }
-    }
+    return (
+        <div className="body" ref={bodyRef}>
+            {
+                bodyData.map((unit) =>
+                    <BodyUnit unit={unit} />
+                )
+            }
+        </div>
+    );
+}
 
-    onDataLoaded() {
-        if (super.onDataLoaded) {
-            super.onDataLoaded();
-        }
-        if (this.el) {
-            this.attachUnits();
+function Tags({tagData}) {
+    return tagData.length > 0 &&
+        <div className="tags">
+            {
+                tagData.map((tag) =>
+                    <div className="tag">{tag}</div>
+                )
+            }
+        </div>
+    ;
+}
+
+function FloatingSideBar({readTime, progress}) {
+    return (
+        <div className="floater">
+            <div className="sticky-bit">
+                <ProgressRing radius="45" progress={progress} stroke="4" message={readTime} />
+                <ShareJsx
+                    pageUrl={encodeURIComponent(window.location.href)}
+                    message={encodeURIComponent('Check out this OpenStax blog article!')}
+                    minimal={true}
+                />
+            </div>
+        </div>
+    );
+}
+
+function useScrollProgress(ref) {
+    const [progress, setProgress] = useState(0);
+    const [bodyRef, setBodyRef] = useState();
+
+    useEffect(() => {
+        if (typeof bodyRef === 'undefined') {
+            return null;
         }
-        const meta = this.pageData.meta;
-    }
+        const handleScroll = throttle(() => setProgress(getProgress(bodyRef.current)));
 
-    onClose() {
-        window.removeEventListener('scroll', this.handleScroll);
-        window.removeEventListener('resize', this.handleScroll);
+        Array.from(ref.current.querySelectorAll('img'))
+            .forEach((img) => {
+                img.onload = handleScroll;
+            });
+        window.addEventListener('scroll', handleScroll);
+        window.addEventListener('resize', handleScroll);
+        handleScroll();
+
+        return () => {
+            window.removeEventListener('scroll', handleScroll);
+            window.removeEventListener('resize', handleScroll);
+        };
+    }, [bodyRef]);
+
+    return [progress, setBodyRef];
+}
```

Comment:
hm, looks like you already have a scroll hook?  in that case I guess maybe use this?

Reply:
I made this use the WindowContext. I also fixed an awkward misunderstanding about forward refs (around `bodyRef`).